# Project Euler Problem 20

n! 이라는 표기법은 n × (n − 1) × ... × 3 × 2 × 1을 뜻합니다.

예를 들자면 10! = 10 × 9 × ... × 3 × 2 × 1 = 3628800 이 되는데,
여기서 10!의 각 자리수를 더해 보면 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27 입니다.

100! 의 자리수를 모두 더하면 얼마입니까?

In [2]:
def cipher_sum(num):
    fac = 1
    each_cipher_sum = 0
    for i in range(1, num+1):
        fac *= i
        
    string_fac = str(fac)
    for j in string_fac:
        each_cipher_sum += int(j)
        
        
    return fac, each_cipher_sum

cases = [4, 5, 10, 100]

for case in cases:
    print cipher_sum(case)

(24, 6)
(120, 3)
(3628800, 27)
(93326215443944152681699238856266700490715968264381621468592963895217599993229915608941463976156518286253697920827223758251185210916864000000000000000000000000L, 648)


* 4!의 경우 4 x 3 x 2 x 1 = 24, 2 + 4 = 6
* 5!의 경우 5 x 4 x 3 x 2 x 1 = 120, 1 + 2 + 0 = 3
* 10!의 경우 10 x 9 x 8 x 7 x ... x 1 = 3628800, 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27

위 세가지의 시도가 모두 맞음.
따라서 100!의 자리수를 모두 더하면 648이다.

